pqarrow/arrowutils/sort.go

```diff
@@ -2,93 +2,306 @@ package arrowutils
+	defer indicesBuilder.Release()
 
 	if r.NumRows() == 0 {
-		return indicesBuilder.NewInt64Array(), nil
+		return indicesBuilder.NewArray().(*array.Int32), nil
 	}
 	if r.NumRows() == 1 {
 		indicesBuilder.Append(0)
-		return indicesBuilder.NewInt64Array(), nil
+		return indicesBuilder.NewArray().(*array.Int32), nil
+	}
+	indicesBuilder.ReserveToLength(int(r.NumRows()))
+	for i := 0; i < int(r.NumRows()); i++ {
+		indicesBuilder.UnsafeSet(i, int32(i))
+	}
+	ms, err := newMultiColSorter(r, indicesBuilder, columns)
+	if err != nil {
+		return nil, err
 	}
+	sort.Sort(ms)
+	return indicesBuilder.NewArray().(*array.Int32), nil
+}
 
-	indices := make([]int64, r.NumRows())
-	// populate indices
-	for i := range indices {
-		indices[i] = int64(i)
+// ReorderRecord uses indices to create a new record that is sorted according to
+// the indices array.
+func ReorderRecord(ctx context.Context, mem memory.Allocator, r arrow.Record, indices *array.Int32) (arrow.Record, error) {
+	ctx = compute.WithAllocator(ctx, mem)
+	// compute.Take doesn't support dictionaries. Use take on r when r does not have
+	// dictionary column.
+	var hasDictionary bool
+	for i := 0; i < int(r.NumCols()); i++ {
+		if r.Column(i).DataType().ID() == arrow.DICTIONARY {
+			hasDictionary = true
+			break
+		}
 	}
+	if !hasDictionary {
+		res, err := compute.Take(
+			ctx,
+			compute.TakeOptions{BoundsCheck: true},
+			compute.NewDatum(r),
+			compute.NewDatum(indices),
+		)
+		if err != nil {
+			return nil, err
+		}
+		return res.(*compute.RecordDatum).Value, nil
+	}
+	resArr := make([]arrow.Array, r.NumCols())
 
-	switch c := r.Column(cols[0]).(type) {
-	case *array.Int64:
-		sort.Sort(orderedSorter[int64]{array: c, indices: indices})
-	case *array.String:
-		sort.Sort(orderedSorter[string]{array: c, indices: indices})
-	default:
-		return nil, fmt.Errorf("unsupported column type for sorting %T", c)
+	defer func() {
+		for _, a := range resArr {
+			if a != nil {
+				a.Release()
+			}
+		}
+	}()
+	var g errgroup.Group
+	for i := 0; i < int(r.NumCols()); i++ {
+		i := i
+		col := r.Column(i)
+		if d, ok := col.(*array.Dictionary); ok {
+			g.Go(takeDictColumn(ctx, d, i, resArr, indices))
+		} else {
+			g.Go(takeColumn(ctx, col, i, resArr, indices))
+		}
+	}
+	err := g.Wait()
+	if err != nil {
+		return nil, err
 	}
+	return array.NewRecord(r.Schema(), resArr, r.NumRows()), nil
+}
 
-	indicesBuilder.Reserve(len(indices))
-	for _, i := range indices {
-		indicesBuilder.Append(i)
+func takeColumn(ctx context.Context, a arrow.Array, idx int, arr []arrow.Array, indices *array.Int32) func() error {
+	return func() error {
+		r, err := compute.TakeArray(ctx, a, indices)
+		if err != nil {
+			return err
+		}
+		arr[idx] = r
+		return nil
 	}
+}
 
-	return indicesBuilder.NewInt64Array(), nil
+func takeDictColumn(ctx context.Context, a *array.Dictionary, idx int, arr []arrow.Array, indices *array.Int32) func() error {
```

Comment:
nit: how about simply returning an error instead of a callback in all these `take` functions?? You'd have to call this within a `func() error { }` block in the `Go` above, but I think it's nicer given that it gives the option to directly execute this function

Reply:
done 7be2e2b